chainer/dataset/convert.py

```diff
@@ -0,0 +1,114 @@
+import numpy
+import six
+
+from chainer import cuda
+
+
+def concat_examples(batch, device=None, padding=None):
+    """Concatenates a list of examples into array(s).
+
+    Dataset iterator yields a list of examples. If each example is an array,
+    this function concatenates them along the newly-inserted first axis (called
+    `batch dimension`) into one array. The basic behavior is same for examples
+    consisting of multiple arrays, i.e., corresponding arrays of all examples
+    are concatenated.
+
+    For instance, consider each example consists of two arrays ``(x, y)``.
+    Then, this function concatenates ``x`` 's into one array, and ``y`` 's
+    into another array, and returns a tuple of these two arrays. Another
+    example: consider each example is a dictionary of two arrays. Two arrays
+    have keys ``'x'`` and ``'y'``. Then, this function concatenates ``x`` 's
+    into one array, and ``y`` 's into another array, and returns a dictionary
+    with two arrays ``x`` and ``y``.
+
+    When the arrays to concatenate have different shapes, the behavior depends
+    on the ``padding`` value. If ``padding`` is None (default), it raises an
+    error. Otherwise, it builds an array of the minimum shape that the contents
+    of all arrays can be substituted to. The padding value is then used to the
+    extra elements of the resulting arrays.
+
+    TODO(beam2d): Add an example.
+
+    Args:
+        batch (list): A list of examples. This is typically given by a dataset
+            iterator.
+        device (int): Device ID to which each array is sent. Negative value
+            indicates the host memory (CPU). If it is omitted, all arrays are
+            left in the original device.
+        padding: Padding value for extra elements. If this is None (default),
+            an error is raised on shape mismatch. Otherwise, an array of
+            minimum dimensionalities that can accomodate all arrays is created,
+            and elements outside of the examples are padded by this value.
+
+    Returns:
+        Array, a tuple of arrays, or a dictionary of arrays. The type depends
+        on the type of each example in the batch.
+
+    """
+    if len(batch) == 0:
+        raise ValueError('batch is empty')
+
+    if device is None:
+        def to_device(x):
+            return x
+    elif device < 0:
+        to_device = cuda.to_cpu
+    else:
+        to_device = lambda x: cuda.to_gpu(x, device)
```

Comment:
There is no unified function that behaves in the same way, thus I had to prepare my own one. Do you think it is better to make it one function? Implication of the current implementation is that it incurs less overheads in the cases where to_device is called many times.